protobuf-format.md

```diff
@@ -0,0 +1,387 @@
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "cloudevents.io/protobuf/";
+option java_package = "io.cloudevents";
+option java_multiple_files = true;
+
+message CloudEvent {
+  string event_type = 1;
+  string event_type_version = 2;
+  string cloud_events_version = 3;
+  string source = 4;
+  string event_id = 5;
+  google.protobuf.Timestamp timestamp = 6;
+  string schema_url = 7;
+  string content_type = 8;
+  // `oneof payload` itself is not a real field, it only enforces the
+  // `oneof` constraint
+  oneof payload {
+    google.protobuf.Value data = 9;
+    bytes bytes_data = 10;
+    google.protobuf.Any proto_data = 11;
+  }
+  google.protobuf.Struct extensions = 12;
+}
+```
+
+In general, the CloudEvents attribute names are converted into
+protobuf fields in lower_snake_case, and the the CloudEvents type
+system is mapped into protobuf types as follows:
+
+
+| CloudEvents | Protobuf
+|--------------|-------------------------------------------------------------
+| String       | string
+| Binary       | bytes
+| URI          | string
+| Timestamp    | google.protobuf.Timestamp
+| Map          | google.protobuf.Struct
+| Object       | google.protobuf.Value
+| Integer | int32
+
+### 2.1 Note on cloud_events_version:
+
+The protobuf `package` keyword defines a package name that includes
+the CloudEvents major version number. The package name and message
+name together form a fully qualified name for a protobuf
+message. Messages with the same name but different package names are
+independent messages with no relation to one another. The
+`cloud_events_version` field’s major version number MUST match the
+major version of the protobuf package.
+
+### 2.2 Special handling of `data` attribute:
+
+#### 2.2.1 Background knowledge on special built in protobuf types
+
+`google.protobuf.Value` ([more info][PROTO_JSON]) is used to express a
+JSON structure. Its fields are composed of special wrapper value types
+such as `google.protobuf.BytesValue`. The protobuf runtime library
+provides JSON conversion support for these special built in message
+types.
+
+`google.protobuf.Any` ([more info][PROTO_ANY]) is used in protobuf to
+express a field whose type is not known to the enclosing message. It
+contains the binary representation of the embedded message and a URL
+string that identifies the type of the embedded message.
+
+
+#### 2.2.2 Handling instructions
+
+If the `contentType` is `application/json` or any media type with the
+structured `+json` suffix, the implementation MUST store the JSON
+payload in the `data` field.
+
+If the payload is binary, the implementation MUST store the bytes in
+the `bytes_data` field.
+
+If the payload is a protobuf, the implementation MUST store the
+payload in the `proto_data` field and the `contentType` MUST be
+`application/protobuf`.
+
+### 2.3 Extensions:
+
+`google.protobuf.Struct` ([more info][PROTO_STRUCT]) represents an
```

Comment:
You could indeed take this construct and move it up one level and make the entire event a key/value pair map. Since you seem to be okay with doing that for extensions, why not also do that for the standard attributes? 

If you're principally worried about wire footprint, you could just use special-prefixed keys like "$0" and "$1" as shorthand for the well-known names. 

The complete wire footprint in proto for "$0", including type identifier and length prefix is 4 bytes, literally "x12 x02 x24 x30". We have 8 well-known attributes right now (with at least 1 of them in doubt and only 4 required). If we'd reach into the ASCII control character range instead of resorting to printable chars, I could cut another byte, e.g. `x12 x01 x01`

When we look at the required attributes, that's between 12 and 16 bytes extra "overhead" compared to pure positional encoding, and that's even within the TLS padding range, i.e. if you save those 12-16 bytes, TLS (the AES 128-bit block cipher inside of it) will pad at least some if not all of those savings away with extra zeroes when you put a single event on the wire. 

The effect of all this extra complication may quite literally come out to nothing.

(Mind that I'm not counting whatever the required overhead for the dictionary is because you've got that anyways for the extensions field defined here) 

I do realize that this might constrain the value type choice for the standard dictionary entries at the tooling level, but this spec seems to be unconcerned about that for extension values, which appear to be encoded as JSON strings.

At the wire level, two lower two bits of the preamble of any encoded protobuf field actually indicates the wire type, which means that the format factually can encode polymorphic dictionaries that distinguish between numbers and strings, which is sufficient for the value space we have in CloudEvents.